File: infrastructures/rest-ddd/typescript/src/infrastructure/repositories/impl/ExampleQueryRepository.ts
/**
 * Example Query Repository Implementation
 * Handles read operations (SELECT).
 */

import { IQueryRepository } from '../IQueryRepository';
import { ExampleReadEntity } from '../../persistence/read/ExampleReadEntity';

export class ExampleQueryRepository implements IQueryRepository<ExampleReadEntity, string> {
  /**
   * Query Repository for Example entity.
   * 
   * Rules:
   * - Located in /infrastructure/repositories/impl/
   * - Suffix with 'Repository'
   * - Returns ReadEntity directly (NO BMO mapping)
   * - Optimized for query performance
   */

  constructor(
    // Add database connection/context here
    // private readonly _dbContext: DatabaseContext
  ) {}

  async findByIdAsync(p_id: string): Promise<ExampleReadEntity | null> {
    // Query database for read entity
    // const result = await this._dbContext.examplesView.findOne({ _id: p_id });
    // 
    // if (!result) {
    //   return null;
    // }
    // 
    // return new ExampleReadEntity(
    //   result._id,
    //   result.name,
    //   result.description,
    //   result.status,
    //   result.ownerId,
    //   result.ownerName,
    //   result.createdAt,
    //   result.updatedAt
    // );

    // Placeholder
    return null;
  }

  async listByFilterAsync(
    p_filter: Record<string, any>,
    p_page: number = 1,
    p_pageSize: number = 50
  ): Promise<ExampleReadEntity[]> {
    // Query database with filter and pagination
    // const skip = (p_page - 1) * p_pageSize;
    // const results = await this._dbContext.examplesView
    //   .find(p_filter)
    //   .skip(skip)
    //   .limit(p_pageSize)
    //   .toArray();
    // 
    // return results.map(r => new ExampleReadEntity(
    //   r._id,
    //   r.name,
    //   r.description,
    //   r.status,
    //   r.ownerId,
    //   r.ownerName,
    //   r.createdAt,
    //   r.updatedAt
    // ));

    // Placeholder
    return [];
  }

  async countByFilterAsync(p_filter: Record<string, any>): Promise<number> {
    // Count documents matching filter
    // return await this._dbContext.examplesView.countDocuments(p_filter);

    // Placeholder
    return 0;
  }
}

File: infrastructures/rest-ddd/csharp/src/Infrastructure/Repositories/Impl/ExampleQueryRepository.cs
/**
 * Example Query Repository Implementation
 * Handles read operations (SELECT).
 */

using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using Infrastructure.Persistence.Read;

namespace Infrastructure.Repositories.Impl
{
    public class ExampleQueryRepository : IQueryRepository<ExampleReadEntity, string>
    {
        /**
         * Query Repository for Example entity.
         * 
         * Rules:
         * - Located in /Infrastructure/Repositories/Impl/
         * - Suffix with 'Repository'
         * - Returns ReadEntity directly (NO BMO mapping)
         * - Optimized for query performance
         */

        // Add database connection/context here
        // private readonly DatabaseContext _dbContext;

        public ExampleQueryRepository()
        {
        }

        public async Task<ExampleReadEntity?> FindByIdAsync(string p_id)
        {
            // Query database for read entity
            // var result = await _dbContext.ExamplesView
            //     .Find(e => e.Id == p_id)
            //     .FirstOrDefaultAsync();
            // 
            // if (result == null)
            // {
            //     return null;
            // }
            // 
            // return new ExampleReadEntity(
            //     result.Id,
            //     result.Name,
            //     result.Description,
            //     result.Status,
            //     result.OwnerId,
            //     result.OwnerName,
            //     result.CreatedAt,
            //     result.UpdatedAt
            // );

            // Placeholder
            await Task.CompletedTask;
            return null;
        }

        public async Task<List<ExampleReadEntity>> ListByFilterAsync(
            Dictionary<string, object> p_filter,
            int p_page = 1,
            int p_pageSize = 50
        )
        {
            // Query database with filter and pagination
            // var skip = (p_page - 1) * p_pageSize;
            // var results = await _dbContext.ExamplesView
            //     .Find(BuildFilter(p_filter))
            //     .Skip(skip)
            //     .Limit(p_pageSize)
            //     .ToListAsync();
            // 
            // return results.Select(r => new ExampleReadEntity(
            //     r.Id,
            //     r.Name,
            //     r.Description,
            //     r.Status,
            //     r.OwnerId,
            //     r.OwnerName,
            //     r.CreatedAt,
            //     r.UpdatedAt
            // )).ToList();

            // Placeholder
            await Task.CompletedTask;
            return new List<ExampleReadEntity>();
        }

        public async Task<int> CountByFilterAsync(Dictionary<string, object> p_filter)
        {
            // Count documents matching filter
            // return await _dbContext.ExamplesView
            //     .CountDocumentsAsync(BuildFilter(p_filter));

            // Placeholder
            await Task.CompletedTask;
            return 0;
        }
    }
}
